Math/NumberTheory/EisensteinIntegers.hs

```diff
@@ -0,0 +1,345 @@
+  , norm
+  , associates
+  , ids
+
+  -- * Division and remainder functions
+  , divE
+  , divModE
+  , modE
+  , quotRemE
+  , quotE
+  , remE
+
+  , gcdE
+
+  -- * Primality functions
+  , factorise
+  , findPrime
+  , isPrime
+  , primes
+  ) where
+
+import Data.List                                       (mapAccumL, partition)
+import Data.Maybe                                      (fromMaybe)
+import Data.Ord                                        (comparing)
+import GHC.Generics                                    (Generic)
+
+import qualified Math.NumberTheory.Moduli               as Moduli
+import Math.NumberTheory.Moduli.Sqrt                    (FieldCharacteristic(..))
+import qualified Math.NumberTheory.Primes.Factorisation as Factorisation
+import Math.NumberTheory.Primes.Types                   (PrimeNat(..))
+import qualified Math.NumberTheory.Primes.Sieve         as Sieve
+import qualified Math.NumberTheory.Primes.Testing       as Testing
+import Math.NumberTheory.Utils                          (mergeBy)
+import Math.NumberTheory.Utils.FromIntegral             (integerToNatural)
+
+infix 6 :+
+
+-- | An Eisenstein integer is a + bω, where a and b are both integers.
+data EisensteinInteger = (:+) { real :: !Integer, imag :: !Integer }
+    deriving (Eq, Ord, Generic)
+
+-- | The imaginary unit for Eisenstein integers, where
+--
+-- > ω == (-1/2) + ((sqrt 3)/2)ι == exp(2*pi*ι/3)
+-- and ι is the usual imaginary unit with ι² == -1.
+ω :: EisensteinInteger
+ω = 0 :+ 1
+
+instance Show EisensteinInteger where
+    show (a :+ b)
+        | b == 0     = show a
+        | a == 0     = s ++ b'
+        | otherwise  = show a ++ op ++ b'
+        where
+            b' = if abs b == 1 then "ω" else show (abs b) ++ "*ω"
+            op = if b > 0 then "+" else "-"
+            s  = if b > 0 then "" else "-"
+
+instance Num EisensteinInteger where
+    (+) (a :+ b) (c :+ d) = (a + c) :+ (b + d)
+    (*) (a :+ b) (c :+ d) = (a * c - b * d) :+ (b * c + a * d - b * d)
+    abs z@(a :+ b)
+        | a == 0 && b == 0                                  = z            -- origin
+        | a > b && b >= 0                                   = z            -- first sextant: 0 ≤ Arg(η) < π/3
+        | b >= a && a > 0                                   = (-ω) * z     -- second sextant: π/3 ≤ Arg(η) < 2π/3
+        | b > 0 && 0 >= a                                   = (-1 - ω) * z -- third sextant: 2π/3 ≤ Arg(η) < π
+        | a < b && b <= 0                                   = - z          -- fourth sextant: -π < Arg(η) < -2π/3 or Arg(η) = π
+        | b <= a && a < 0                                   = ω * z        -- fifth sextant: -2π/3 ≤ Arg(η) < -π/3
+        | otherwise                                         = (1 + ω) * z  -- sixth sextant: -π/3 ≤ Arg(η) < 0
+    negate (a :+ b) = (-a) :+ (-b)
+    fromInteger n = n :+ 0
+    signum z@(a :+ b)
+        | a == 0 && b == 0 = z               -- hole at origin
+        | otherwise        = z `divE` abs z
```

Comment:
It would be nice to avoid division here (and for Gaussian integers as well, generally speaking). To do not repeat a long pattern-matching we can have a single function `absSignum :: EisensteinInteger -> (EisensteinInteger, EisensteinInteger)` and `abs = fst . absSignum`, `signum = snd . absSignum`.